File: src/main/java/com/yuyan/imemodule/utils/expression/ExpressionBuilder.java
package com.yuyan.imemodule.utils.expression;

import com.yuyan.imemodule.utils.expression.function.Function;
import com.yuyan.imemodule.utils.expression.function.Functions;
import com.yuyan.imemodule.utils.expression.operator.Operator;
import com.yuyan.imemodule.utils.expression.shuntingyard.ShuntingYard;

import java.util.*;

/**
 * Factory class for {@link Expression} instances. This class is the main API entrypoint. Users should create new
 * {@link Expression} instances using this factory class.
 */
public class ExpressionBuilder {

    private final String expression;

    private final Map<String, Function> userFunctions;

    private final Map<String, Operator> userOperators;

    private final Set<String> variableNames;

    private boolean implicitMultiplication = true;

    /**
     * Create a new ExpressionBuilder instance and initialize it with a given expression string.
     *
     * @param expression the expression to be parsed
     */
    public ExpressionBuilder(String expression) {
        if (expression == null || expression.trim().length() == 0) {
            throw new IllegalArgumentException("Expression can not be empty");
        }
        this.expression = expression;
        this.userOperators = new HashMap<>(4);
        this.userFunctions = new HashMap<>(4);
        this.variableNames = new HashSet<>(4);
    }

    /**
     * Build the {@link Expression} instance using the custom operators and functions set.
     *
     * @return an {@link Expression} instance which can be used to evaluate the result of the expression
     */
    public Expression build() {
        if (expression.length() == 0) {
            throw new IllegalArgumentException("The expression can not be empty");
        }

        /* set the constants' varibale names */
        variableNames.add("pi");
        variableNames.add("π");
        variableNames.add("e");
        variableNames.add("φ");

        /* Check if there are duplicate vars/functions */
        for (String var : variableNames) {
            if (Functions.getBuiltinFunction(var) != null || userFunctions.containsKey(var)) {
                throw new IllegalArgumentException("A variable can not have the same name as a function [" + var + "]");
            }
        }

        return new Expression(ShuntingYard.convertToRPN(this.expression, this.userFunctions, this.userOperators,
                this.variableNames, this.implicitMultiplication), this.userFunctions.keySet());
    }

}

File: src/main/java/com/yuyan/imemodule/utils/expression/ExpressionBuilder.kt
package com.yuyan.imemodule.utils.expression

import com.yuyan.imemodule.utils.expression.function.Function
import com.yuyan.imemodule.utils.expression.function.Functions
import com.yuyan.imemodule.utils.expression.operator.Operator
import com.yuyan.imemodule.utils.expression.shuntingyard.ShuntingYard

/**
 * Factory class for [Expression] instances. This class is the main API entrypoint. Users should create new
 * [Expression] instances using this factory class.
 */
class ExpressionBuilder(expression: String?) {
    private val expression: String
    private val userFunctions: Map<String, Function>
    private val userOperators: Map<String, Operator>
    private val variableNames: MutableSet<String>
    private val implicitMultiplication = true

    init {
        require(!(expression == null || expression.trim { it <= ' ' }.isEmpty())) { "Expression can not be empty" }
        this.expression = expression
        userOperators = HashMap(4)
        userFunctions = HashMap(4)
        variableNames = HashSet(4)
    }

    /**
     * Build the [Expression] instance using the custom operators and functions set.
     *
     * @return an [Expression] instance which can be used to evaluate the result of the expression
     */
    fun build(): Expression {
        require(expression.isNotEmpty()) { "The expression can not be empty" }
        /* set the constants' varibale names */variableNames.add("pi")
        variableNames.add("π")
        variableNames.add("e")
        variableNames.add("φ")

        /* Check if there are duplicate vars/functions */for (`var` in variableNames) {
            require(!(Functions.getBuiltinFunction(`var`) != null || userFunctions.containsKey(`var`))) { "A variable can not have the same name as a function [$`var`]" }
        }
        return Expression(
            ShuntingYard.convertToRPN(
                expression, userFunctions, userOperators,
                variableNames, implicitMultiplication
            ), userFunctions.keys
        )
    }
}
